task3/index.js

```diff
@@ -0,0 +1,168 @@
+- updateStudentByName(new Student(...), name) - знайти об'єкт студента по name та замінити на student (новий екземпляр ФК Student) */
+
+/* Підсказки та правила, яких потрібно дотримуватися при виконанні завдання
+
+- Всі "діти" конструктори повинні успадковувати getFullName та setFullName.
+- Всі "діти" конструктори не повинні дублювати створення полів у конструкторі.Позбавтеся дублювання за допомогою заміни контексту.
+- Всі конструктори приймають об'єкт у аргументи, а не аргументи через кому
+- Методи описуємо у прототипі
+- Слідкуйте, щоб ваш код не повторювався - якщо потрібно створюйте додаткові методи і не забувайте користуватися готовими.
+- Перевіряйте ваш код декілька разів
+- Читайте уважно умову завдання.
+*/
+
+function Human({ name, surname, age }) {
+	this.name = name;
+	this.surname = surname;
+	this.age = age;
+}
+
+Human.prototype.getFullName = function () {
+	return `${this.name} ${this.surname}`;
+};
+Human.prototype.setFullName = function (fullName) {
+	return ([this.name, this.surname] = fullName.split(' '));
+};
+
+Human.prototype.constructor = Human;
+/*  */
+const user1 = new Human({ name: 'Vladyslav', surname: 'Tsarik', age: 26 });
+
+console.dir(user1);
+
+/* 2) */
+
+function Student({ name, surname, age, mark }) {
+	Human.call(this, { name, surname });
+	this.age = age;
+	this.mark = mark;
+
+	Student.prototype.getAverageMark = function () {
+		return this.mark.reduce((a, b) => a + b) / mark.length;
+	};
+	Student.prototype.getMinMark = function () {
+		return Math.min(...this.mark);
+	};
+	Student.prototype.getMaxMark = function () {
+		return Math.max(...mark);
+	};
+}
+
+/*  */
+const Student1 = new Student({ name: 'John', surname: 'Black', age: 25, mark: [3, 3, 4, 5, 3, 4, 5, 3] });
+const Student2 = new Student({ name: 'Bob', surname: 'White', age: 23, mark: [5, 4, 4, 4, 4, 4, 4, 3] });
+const Student3 = new Student({ name: 'Willyam', surname: 'Skoll', age: 21, mark: [3, 3, 4, 3, 5, 3, 4, 3] });
+const Student4 = new Student({ name: 'Jayne', surname: 'Doe', age: 22, mark: [5, 5, 5, 5, 4, 5, 5, 3] });
+const Student5 = new Student({ name: 'Kate', surname: 'Klime', age: 24, mark: [5, 5, 5, 5, 5, 5, 5, 4] });
+/*  */
+console.dir(Student1);
+console.dir(Student2);
+console.dir(Student3);
+console.dir(Student4);
+console.dir(Student5);
+console.log(Student1.getAverageMark());
+console.log(Student1.getMinMark());
+console.log(Student1.getMaxMark());
+
+/* 3) */
+function Teacher({ name, surname, age, students }) {
+	Human.call(this, { name, surname });
+	this.age = age;
+	this.students = students;
+	/* Methods */
+	Teacher.prototype.getListOfNamesByAverageMark = function () {
+		const studentAvarageMark = this.students
+			.map((item) => {
+				item.averageMark = item.getAverageMark();
+				return item;
+			})
+			.sort((a, b) => {
+				return b.averageMark - a.averageMark;
+			});
+
+		return studentAvarageMark.map((item) => {
+			delete item.averageMark;
+			return item.name;
+		});
```

Comment:
it can be simplified
```suggestion
	        const copiedStudents = [...this.students];
  
                return copiedStudents
                    .sort((a, b) => b.getAverageMark() - a.getAverageMark())
                    .map(student => student.name);
```
now explain me why I need copiedStudents variable?